cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,882 @@
+    handle.get_thrust_policy(),
+    prefix_sum.begin(),
+    prefix_sum.begin() + invalid_dst.size(),
+    [num_edges, invalid_dst, edgelist_dsts = edgelist_dsts.begin()] __device__(auto idx) {
+      auto itr_lower_valid = thrust::lower_bound(
+        thrust::seq, edgelist_dsts, edgelist_dsts + num_edges, invalid_dst[idx]);
+      
+      auto itr_upper_valid = thrust::upper_bound(
+        thrust::seq, itr_lower_valid, edgelist_dsts + num_edges, invalid_dst[idx]);
+
+      auto dist = thrust::distance(itr_lower_valid, itr_upper_valid);
+
+      return dist;
+    });
+  thrust::exclusive_scan(
+    handle.get_thrust_policy(), prefix_sum.begin(), prefix_sum.end(), prefix_sum.begin());
+
+  return prefix_sum;
+}
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+edge_t remove_overcompensating_edges(raft::handle_t const& handle,
+                                     edge_t num_valid_edges,
+                                     edge_t num_invalid_edges,
+                                     edge_t dataframe_buffer_size,
+                                     EdgeIterator&& potential_closing_or_incoming_edges,
+                                     EdgeIterator&& incoming_or_potential_closing_edges,
+                                     raft::device_span<vertex_t> edgelist_srcs,
+                                     raft::device_span<vertex_t> edgelist_dsts)
+{
+  // To avoid over-compensating, check whether the 'potential_closing_edges'
+  // are within the invalid edges. If yes, the was already unrolled
+  auto edges_not_overcomp = thrust::remove_if(
+    handle.get_thrust_policy(),
+    thrust::make_zip_iterator(get_dataframe_buffer_begin(potential_closing_or_incoming_edges),
+                              get_dataframe_buffer_begin(incoming_or_potential_closing_edges)),
+    thrust::make_zip_iterator(
+      get_dataframe_buffer_begin(potential_closing_or_incoming_edges) + dataframe_buffer_size,
+      get_dataframe_buffer_begin(incoming_or_potential_closing_edges) + dataframe_buffer_size),
+    [num_invalid_edges,
+     num_valid_edges,
+     invalid_first = thrust::make_zip_iterator(edgelist_dsts.begin() + num_valid_edges,
+                                               edgelist_srcs.begin() + num_valid_edges),
+     invalid_last =
+       thrust::make_zip_iterator(edgelist_dsts.end(), edgelist_srcs.end())] __device__(auto e) {
+      auto potential_edge = thrust::get<0>(e);
+      auto transposed_potential_or_incoming_edge =
+        thrust::make_tuple(thrust::get<1>(potential_edge), thrust::get<0>(potential_edge));
+      auto itr = thrust::lower_bound(
+        thrust::seq,
+        invalid_first,
+        invalid_last,
+        transposed_potential_or_incoming_edge); 
+  
+      assert(*itr == transposed_potential_or_incoming_edge);
+      auto dist = thrust::distance(invalid_first, itr);
+      if (*itr != transposed_potential_or_incoming_edge) { return false; }
+      return dist < num_invalid_edges;
```

Comment:
`*itr` can lead to illegal memory access if itr == invalid_last

Reply:
`return (itr != invalid_last && *itr == transposed_potential_or_incoming_edges);`

Reply:
Right . No need to check the distance. I can just compare the two edges iff there is no illegal memory access